azavea/stac4s
# Get publishing to Sonatype working

## Overview

The following changes were made in an effort to get publishing to Sonatype working:

- Add secret environment variables to CircleCI using [contexts](https://circleci.com/docs/2.0/contexts/)
- Version artifacts with [sbt-git](https://github.com/sbt/sbt-git)
- Add `publishSettings` to `core` project
- Move dependencies to `Dependencies.scala` to align with other Scala project organization and to fix failing tests due to `Versions.ScapegoatVersion` reference error
- Remove `onLoad` setting and aggregate `core` under a root project instead so that `scalafmtSbtCheck` has visibility into top-level `*.sbt` and `project/*.scala` files

Connects https://github.com/azavea/raster-foundry-platform/issues/899

## Testing Instructions

See CircleCI build: https://app.circleci.com/jobs/github/azavea/stac4s/13

Retrieve the Azavea public key:

```bash
$ gpg --recv-keys 0x713F9F29598CFFF3
gpg: directory '/home/rocky/.gnupg' created
gpg: keybox '/home/rocky/.gnupg/pubring.kbx' created
gpgpg: /home/rocky/.gnupg/trustdb.gpg: trustdb created
gpg: key 713F9F29598CFFF3: public key "Azavea Inc. <systems@azavea.com>" imported
gpg: Total number processed: 1
gpg:               imported: 1
```

Verify the signature of the published artifact:

```bash
$ curl -O https://oss.sonatype.org/content/repositories/snapshots/com/azavea/stac4s/core_2.12/ba29caf-SNAPSHOT/core_2.12-ba29caf-SNAPSHOT.jar
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  322k  100  322k    0     0  1772k      0 --:--:-- --:--:-- --:--:-- 1772k
$ curl -O https://oss.sonatype.org/content/repositories/snapshots/com/azavea/stac4s/core_2.12/ba29caf-SNAPSHOT/core_2.12-ba29caf-SNAPSHOT.jar.asc
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   833  100   833    0     0  10412      0 --:--:-- --:--:-- --:--:-- 10412
```
```
$ gpg --verify core_2.12-ba29caf-SNAPSHOT.jar.asc core_2.12-ba29caf-SNAPSHOT.jar
gpg: Signature made Wed 11 Dec 2019 12:54:10 PM EST
gpg:                using RSA key 2A780CD99D6FD6711BDD70623C45FEB5E8460BAD
gpg: Good signature from "Azavea Inc. <systems@azavea.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: FD92 9B9D 1047 3BD1 7B56  A718 713F 9F29 598C FFF3
     Subkey fingerprint: 2A78 0CD9 9D6F D671 1BDD  7062 3C45 FEB5 E846 0BAD
```
